### PR TITLE
feat: create new GitHub Authentication Provider extension

### LINKED ISSUE #4079: github API rate limit exceeded
### Bug description

On certain entreprise network, we can get easily the `API rate limit exceeded` github problem.

The following extensions will not work
- kind 

> main ↪️ [kind] Error: API rate limit exceeded for 91.103.43.254. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.)

- compose

> nothing will happen when clicking on the `compose` bottom menu item.

### Operating system

Windows 10

### Installation Method

Installer from website/GitHub releases

### Version

1.4.0

### Steps to reproduce

_No response_

### Relevant log output

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by axel7083**
### What does this PR do?

Adding a GitHub Authentication Provider Extension to fix the rate limit when using GitHub api.

### Screenshot / video of UI

<details>
  <summary>Screenshots</summary>
 

**preference configuration**
image
**sign in**
image
**LOGGED IN**
image 

**configured invalid token on startup**
image

</details>

### What issues does this PR fix or reference?

Closing #4079

### How to test this PR?

- Create an access token
- Put the access token in the properties
- Sign in through the Authentication page

### CODE REVIEW

**Comment by axel7083** [comment]
@benoitf The SonarCloud is giving an error for the build.js script located in the build folder for the extension created, it is a copy/past of the script used in the other extension, any clue ?

**Comment by benoitf** [comment]
@axel7083 we need to review in SonarCloud UI the issues and we can flag them as 'safe' (which is the case there)

I've applied the change

File: extensions/compose/src/utils.ts [hunk 1]
```diff
@@ -26,3 +28,17 @@ export async function makeExecutable(filePath: string): Promise<void> {
     await promises.chmod(filePath, 0o755);
   }
 }
+
+export async function createOctokitClient(): Promise<Octokit> {
+  console.log('kind createOctokitClient');
```
Comment (lstocchi) [inline extensions/compose/src/utils.ts]: to be removed?

File: extensions/kind/src/util.ts [hunk 2]
```diff
@@ -158,3 +159,17 @@ export async function getMemTotalInfo(socketPath: string): Promise<number> {
     });
   });
 }
+
+export async function createOctokitClient(): Promise<Octokit> {
+  console.log('kind createOctokitClient');
```
Comment (lstocchi) [inline extensions/kind/src/util.ts]: to be removed?

**Review by lstocchi — APPROVED** [review]
LGTM

File: extensions/github/src/extension.ts [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    this._onDidChangeSession.fire({
+      removed: removed,
+    });
+    this.sessions = this.sessions.filter(session => session.id !== sessionId);
+  }
+}
+
+let disposable: extensionApi.Disposable | undefined = undefined;
+
+export async function activate(): Promise<void> {
```
Comment (dgolovin) [inline extensions/github/src/extension.ts]:
```suggestion
export async function activate(extensionContext: extensionApi.ExtensionContext): Promise<void> {
```

File: extensions/github/src/extension.ts [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+  async removeSession(sessionId: string): Promise<void> {
+    const removed = this.sessions.filter(session => session.id === sessionId);
+    this._onDidChangeSession.fire({
+      removed: removed,
+    });
+    this.sessions = this.sessions.filter(session => session.id !== sessionId);
+  }
+}
+
+let disposable: extensionApi.Disposable | undefined = undefined;
```
Comment (dgolovin) [inline extensions/github/src/extension.ts]:
```suggestion
```

File: extensions/github/src/extension.ts [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+export async function activate(): Promise<void> {
+  const telemetryLogger = extensionApi.env.createTelemetryLogger();
+  const authProvider = new GitHubProvider(telemetryLogger);
+
+  // Register the AuthenticationProvider
+  disposable = extensionApi.authentication.registerAuthenticationProvider('github', 'GitHub', authProvider, {
```
Comment (dgolovin) [inline extensions/github/src/extension.ts]:
```suggestion
  const disposable = extensionApi.authentication.registerAuthenticationProvider('github', 'GitHub', authProvider, {
```

File: extensions/github/src/extension.ts [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+        light: './logo-light.png',
+        dark: './logo-dark.png',
+      },
+      logo: {
+        light: './logo-light.png',
+        dark: './logo-dark.png',
+      },
+    },
+  });
+
```
Comment (dgolovin) [inline extensions/github/src/extension.ts]:
```suggestion
extensionContext.subscriptions.push(disposable);
```

File: extensions/github/src/extension.ts [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+      await authProvider.createSession([]);
+    } catch (e) {
+      await window.showErrorMessage(`GitHub access token is not valid.`);
+      console.error(e);
+    }
+  }
+}
+
+export function deactivate(): void {
+  disposable?.dispose();
```
Comment (dgolovin) [inline extensions/github/src/extension.ts]:
```suggestion
   console.log('stopping github extension');
```

**Review by dgolovin — CHANGES_REQUESTED** [review]
1. There is not need for global var to store disposable, just use extensionContext.subscriptions
2. Refactoring for king and compose should be in additional PR.

**Comment by axel7083** [comment]
> 1. There is not need for global var to store disposable, just use extensionContext.subscriptions
Thanks for the tip! 

> 2. Refactoring for king and compose should be in additional PR.

I created a separate branch https://github.com/axel7083/podman-desktop/tree/refactor/kind-compose-github-provider I will open a PR when this one this be merged

CC @dgolovin

File: extensions/github/src/extension.ts [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+  // Register the AuthenticationProvider
+  const disposable = extensionApi.authentication.registerAuthenticationProvider('github', 'GitHub', authProvider, {
+    images: {
+      icon: {
+        light: './logo-light.png',
+        dark: './logo-dark.png',
+      },
+      logo: {
+        light: './logo-light.png',
+        dark: './logo-dark.png',
```
Comment (benoitf) [inline extensions/github/src/extension.ts]: it looks light and dark icon are the same

```
d83ddf4ea98e9d2e77848869cbe5f0ecabb3058ede22c681bbe07b45a9091909 logo-light.png
d83ddf4ea98e9d2e77848869cbe5f0ecabb3058ede22c681bbe07b45a9091909 logo-dark.png
```

so either we have different images or we just provide the icon with the simple form

```ts
icon: 'icon.png'
```

File: extensions/github/tsconfig.json [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+    "rootDir": "src",
+    "outDir": "dist",
+    "skipLibCheck": true,
+    "types": ["node"]
+  },
+  "include": ["src", "types/*.d.ts", "../../types/**/*.d.ts"],
+  "ts-node": {
+    "compilerOptions": {
+      "module": "CommonJS",
+      "lib": ["ES2020", "DOM"],
```
Comment (benoitf) [inline extensions/github/tsconfig.json]: not sure DOM should be present there but I suppose it's already somewhere in the code that you copy/paste

File: extensions/github/src/extension.ts [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+  });
+  extensionContext.subscriptions.push(disposable);
+  // If we have a defined GitHub preference access token, let's create the session
+  if (getGitHubAccessToken()) {
+    try {
+      await authProvider.createSession([]);
+    } catch (e) {
+      await window.showErrorMessage(`GitHub access token is not valid.`);
+      console.error(e);
+    }
```
Comment (benoitf) [inline extensions/github/src/extension.ts]: it looks like we're not testing this part.
Should the error message also be displayed to the user ?

File: extensions/github/src/extension.ts [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+        light: './logo-light.png',
+        dark: './logo-dark.png',
+      },
+    },
+  });
+  extensionContext.subscriptions.push(disposable);
+  // If we have a defined GitHub preference access token, let's create the session
+  if (getGitHubAccessToken()) {
+    try {
+      await authProvider.createSession([]);
```
Comment (benoitf) [inline extensions/github/src/extension.ts:131]: it does not need any scopes ?

File: package.json [hunk 12]
```diff
@@ -23,12 +23,13 @@
...
+    "build:extensions": "npm run build:extensions:compose && npm run build:extensions:github && npm run build:extensions:docker && npm run build:extensions:lima && npm run build:extensions:podman && npm run build:extensions:kubecontext && npm run build:extensions:kind && npm run build:extensions:registries && npm run build:extensions:kubectl-cli",
     "build:extensions:compose": "cd ./extensions/compose && npm run build",
     "build:extensions:docker": "cd ./extensions/docker && npm run build",
     "build:extensions:kubecontext": "cd ./extensions/kube-context && npm run build",
     "build:extensions:kind": "cd ./extensions/kind && npm run build",
     "build:extensions:lima": "cd ./extensions/lima && npm run build",
+    "build:extensions:github": "cd ./extensions/github && npm run build",
     "build:extensions:podman": "cd ./extensions/podman && npm run build",
     "build:extensions:registries": "cd ./extensions/registries && npm run build",
```
Comment (benoitf) [inline package.json:34]: `test:extensions` task needs to be enhanced as well to include the tests of this extension

b/c for example for now the tests are not passing but we don't see it

**Review by benoitf — CHANGES_REQUESTED** [review]

File: extensions/github/src/extension.ts (hunk 11 above)
Comment (axel7083) [reply extensions/github/src/extension.ts:131]: I do not see we could scope it ?

File: package.json (hunk 12 above)
Comment (axel7083) [reply package.json:34]: Fixed

File: extensions/github/src/extension.ts (hunk 10 above)
Comment (axel7083) [reply extensions/github/src/extension.ts]: It is tested in https://github.com/containers/podman-desktop/blob/2539606fc9a13e476fa208648d4b838e5fe640f4/extensions/github/src/extension.spec.ts#L100

File: extensions/github/tsconfig.json (hunk 9 above)
Comment (axel7083) [reply extensions/github/tsconfig.json]: removed

File: extensions/github/src/extension.ts (hunk 8 above)
Comment (axel7083) [reply extensions/github/src/extension.ts]: applied

File: extensions/github/src/extension.ts [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+    }
+  }
+}
+
+export function deactivate(): void {
+  console.log('stopping github extension');
+}
+
+export function getGitHubAccessToken(): string | undefined {
+  return extensionApi.configuration.getConfiguration('github').get('token');
```
Comment (dgolovin) [inline extensions/github/src/extension.ts:144]: A github token should be treated the same way as username/password pair. It cannot be stored in configuration.

**Review by dgolovin — CHANGES_REQUESTED** [review]

**Comment by odockal** [comment]
@axel7083 Hey, I have prepared an epic to cover this feature (https://github.com/containers/podman-desktop/issues/5488), it would be nice if (I haven't fully checked every comment) we have acceptance criteria of this feature, so we are sure to cover everything with e2e tests in future. thx!

File: extensions/github/src/extension.ts (hunk 13 above)
Comment (axel7083) [reply extensions/github/src/extension.ts:144]: @benoitf do you have an idea on the way to proceed because I am a bit lost

Comment (dgolovin) [reply extensions/github/src/extension.ts:144]: Extension context should have `secrets` property to store secrets and that should be encrypted storage.
`safeStorage` and `safeStorageElectron` in electron that can help with encryption.

Example how to use it from VS Code https://vscode.dev/github/microsoft/vscode/blob/main/src/vs/platform/encryption/electron-main/encryptionMainService.ts#L32

**Comment by jeffmaury** [comment]
@axel7083 should we keep it ?

**Comment by axel7083** [comment]
> @axel7083 should we keep it ?

Closing it as I totally forgot about it

**Pull request closed without merging**